### PR TITLE
Add ObjCStrInstance and disable automatic NSString-to-str conversion (fixes #101)

### LINKED ISSUE #101: Remove automatic NSString-to-str conversion
As [discussed on Gitter](https://gitter.im/pybee/general?at=5a767662e217167e2c3c2d53), this behavior is problematic when you want to create a new `NSString` using a factory or init method, or when you want to call methods on a `NSString` returned from a method.

Instead, an `ObjCStringInstance` class should be added, which adds a Pythonic interface to `NSString` objects.

### PR DESCRIPTION
**Pull request by dgelessus**
This disables the automatic conversion of `NSString` objects to `str` when returned from Objective-C methods, as discussed in #101. As a replacement, it adds an internal `ObjCStrInstance` class, which provides Pythonic operators and methods on `NSString` objects. (This is similar to what `ObjCListInstance` does for `NSArray` objects.)

There are a few small differences between "real" `str`s and `ObjCStrInstance` objects:

* `str`s consist of Unicode code points, whereas `NSString`s consist of UTF-16 code units. (See the `ObjCStrInstance` docstring for a longer explanation.) I've chosen to use the `NSString` behavior for `ObjCStrInstance`, because attempting to emulate the Python `str` behavior would be very complicated and error-prone. In practice the difference rarely matters, and when it does, the `NSString` can be converted to a `str`.
* `ObjCStrInstance` does not implement all the operations that `str` supports, only the most common ones (indexing, slicing, `+`, `*`, `in`, `find`, `index`). It might be useful to implement some additional methods in the future, such as `replace`, `partition`, `split`, and `join`. Others aren't worth reimplementing, either because they are too complex (`format`, `encode`) or too rarely used (`center`, `expandtabs`). If one of the missing methods is needed, the `NSString` can be converted to a `str`.

### CODE REVIEW

**Comment by freakboy3742** [comment]
@dgelessus Just letting you know I haven't forgotten about this; I've just been snowed under with work and PyCon US preparations. I'm hoping to take a look this weekend.

File: rubicon/objc/runtime.py [hunk 1]
```diff
@@ -1739,8 +1734,10 @@ def py_from_ns(nsobj, *, _auto=False):
     elif _auto:
         # _auto is a private kwarg that is only passed when py_from_ns is called to perform an implicit conversion
         # of a method return value or argument into Python. In this case we only want to perform a few simple
-        # conversions (strings and numbers).
+        # conversions (boxed numbers to their Python equivalents).
         return nsobj
+    elif nsobj.isKindOfClass(NSString):
+        return str(nsobj)
```
Comment (freakboy3742) [inline rubicon/objc/runtime.py]: It took me a little bit to work out the significance of this change. It would be helpful to drop a comment noting that everything below the `elif _auto` are the data types that will be converted if you explicitly request conversion to Python, but won't be as part of an automated conversion. The existence of that "_auto" path wasn't a new addition, but it wasn't immediately obvious to me reviewing this, so since you're in the area, might as well clarify for next time.

File: tests/test_NSDictionary.py [hunk 2]
```diff
@@ -51,10 +51,10 @@ def test_getitem(self):
     def test_iter(self):
         d = self.make_dictionary(self.py_dict)
 
-        keys = set(self.py_dict)
+        keys = {str(k) for k in self.py_dict}
         for k in d:
-            self.assertTrue(k in keys)
-            keys.remove(k)
+            self.assertTrue(str(k) in keys)
+            keys.remove(str(k))
```
Comment (freakboy3742) [inline tests/test_NSDictionary.py]: I get why this change is needed from an API point of view, but it feels like it's moved from being a check of NSString behavior, to a test of Python str behavior...

File: tests/test_NSDictionary.py [hunk 3]
```diff
@@ -104,17 +104,20 @@ def test_equivalence(self):
 
     def test_keys(self):
         a = self.make_dictionary(self.py_dict)
-        for k1, k2 in zip(sorted(a.keys()), sorted(self.py_dict.keys())):
+        for k1, k2 in zip(sorted(a.keys(), key=str), sorted(self.py_dict.keys())):
```
[the remark this answers is not in the document]
Comment (dgelessus) [reply tests/test_NSDictionary.py]: Right, for some reason I didn't add ordering support to `ObjCStrInstance`. Now that I think about it, I'm not sure why. 😄 I'll try implementing the ordering operators and see if I run into any issues.

File: tests/test_NSDictionary.py (hunk 2 above)
Comment (dgelessus) [reply tests/test_NSDictionary.py]: I don't think it was ever meant as a test of `NSString` behavior... if it is, then it should be moved into `test_NSString`.

As I understand it, this test checks that `ObjCDictInstance`s can be iterated over correctly. To do that, it copies the Python dict's keys into a (mutable) set, then iterates over the ObjC dict and checks that every key appears exactly once in the iteration. If an incorrect or duplicate key appeared in the iteration, the `in` check would fail; if a key didn't appear in the iteration, the final `len` check would fail.

File: rubicon/objc/collections.py [hunk 4]
```diff
@@ -1,10 +1,159 @@
...
+            return super().__eq__(other)
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    # Note: We cannot define a __hash__ for NSString objects; doing so would violate the Python convention that
+    # mutable objects should not be hashable. Although we could disallow hashing for NSMutableString objects, this
+    # would make some immutable strings unhashable as well, because immutable strings can have a runtime class that
+    # is a subclass of NSMutableString. This is not just a theoretical possibility - for example, on OS X 10.11,
+    # isinstance(NSString.string(), NSMutableString) is true.
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply rubicon/objc/collections.py]: Ok - We should probably document that limitation (and the 'cast to Python' workaround) somewhere to avoid confusion.

File: tests/test_NSDictionary.py (hunk 2 above)
Comment (freakboy3742) [reply tests/test_NSDictionary.py]: The new iteration of this seems a more direct test of that behavior.

File: rubicon/objc/collections.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dgelessus) [reply rubicon/objc/collections.py]: We need to document more things in general 😃 I'm actually not quite sure what the best place for this information would be. I suppose I could put it in the type conversion howto, since `ObjCDictInstance` and `ObjCListInstance` are explained there too, but I feel like this sort of thing would be better suited for the reference documentation.

I don't think the behavior is that strange either, since normal `ObjCInstance` and all other subclasses are also unhashable. On the other hand, I can see why one might expect it to work, since strings work fine as dictionary keys in pure Python and pure ObjC.

[the remark this answers is not in the document]
Comment (dgelessus) [reply rubicon/objc/collections.py]: OK, done. Part of the new "Strings" section in the docs is copy-pasted from the "Lists" section below (with minor modifications), but I think that's okay for now. I was planning to work on the docs next anyway, so when I do that I'll see if I can rewrite the "Strings", "Lists" and "Dictionaries" sections a bit so they don't repeat themselves as much.
